pkg/endpoint/policy.go

```diff
@@ -258,15 +249,46 @@ func (e *Endpoint) regeneratePolicy(stats *regenerationStatistics) (*policyGener
 	}
 	repo.RUnlock() // Done with policy repository; release this now as Consume() can be slow
 
+	// Add new redirects before Consume() so that all required proxy ports are available for it.
+	var desiredRedirects map[string]uint16
+	err = e.rlockAlive()
+	if err != nil {
+		return nil, err
+	}
+	// Ingress endpoint needs no redirects
+	if !e.isProperty(PropertySkipBPFPolicy) {
+		stats.proxyConfiguration.Start()
+		desiredRedirects, ff, rf = e.addNewRedirects(result.selectorPolicy, datapathRegenCtxt.proxyWaitGroup)
```

Comment:
The vague distinction between `e.regeneratePolicy()` and `e.setDesiredPolicy()` was that `regeneratePolicy()` shouldn't mutate the endpoint, since it drops the endpoint lock for all of the heavy lifting.

I see the issue, though -- we need to know the `proxyPort` to generate the MapState.

Let me think if there's a reasonably clever way to work around this. If not, then c'est la vie.

Reply:
There is a bigger refactor change endpoints to share redirects for the same proxy type and port (underlying proxy listener is already shared), this would move redirect management to policy ingestion and we could fail out on policy ingestion if proxy redirect for it can not be created. One significant obstacle right now is that `pkg/proxy` imports `pkg/policy` and hence policy package can not use the types or functions of the proxy package, at least not directly.

Then, from the viewpoint of the endpoint, if the policy needs redirects, they would already exist and the policy package could place in the right proxy ports without the endpoint package being involved.

However, this is out of the scope for this PR, IMO.

Reply:
This is seemingly blocked by the named port being per-endpoint, but that can be likely separated from the *creation* of the redirect, if per-endpoint policy applicable to the redirects is configured separately.

Reply:
The reality is that calculating policy is highly side-effecting anyways, so the distinction is extremely academic anyways. As long as we hold the endpoint lock for as short a time as possible, I’m satisfied.

Reply:
So we're all good for this PR then?